src/deltalake_ext.rs

```diff
@@ -0,0 +1,322 @@
+        source: StorageError 
+    },
+
+    #[error("DeltaTable interaction failed: {source}")]
+    DeltaTable { 
+        #[from]
+        source: DeltaTableError 
+    },
+
+    #[error("Arrow interaction failed: {source}")]
+    Arrow { 
+        #[from]
+        source: ArrowError 
+    },
+
+    #[error("Parquet write failed: {source}")]
+    Parquet { 
+        #[from]
+        source: ParquetError 
+    },
+}
+
+pub struct DeltaParquetWriter {
+    table_path: String,
+    storage: Box<dyn StorageBackend>,
+    arrow_schema_ref: Arc<arrow::datatypes::Schema>,
+    writer_properties: WriterProperties,
+    cursor: InMemoryWriteableCursor,
+    arrow_writer: ArrowWriter<InMemoryWriteableCursor>,
+    partition_values: HashMap<String, String>,
+    num_records: i64,
+}
+
+/// A writer that writes record batches in parquet format to a table location.
+/// This should be used along side a DeltaTransaction wrapping the same DeltaTable instance.
+impl DeltaParquetWriter {
+    /// Initialize the writer from the given table path and delta schema
+    pub async fn for_table_path_with_schema(table_path: String, schema: &Schema) -> Result<DeltaParquetWriter, DeltaWriterError> {
+        // Initialize storage backend for the given table path
+        let storage = deltalake::get_backend_for_uri(table_path.as_str())?;
+
+        // Initialize an arrow schema ref from the delta table schema
+        let arrow_schema = <ArrowSchema as From<&Schema>>::from(schema);
+        let arrow_schema_ref = Arc::new(arrow_schema);
+
+        // Initialize writer properties for the underlying arrow writer
+        let writer_properties = WriterProperties::builder()
+            // TODO: Extract config/env for writer properties and set more than just compression
+            .set_compression(Compression::SNAPPY)
+            .build();
+
+        let cursor = InMemoryWriteableCursor::default();
+        // TODO: Handle error
+        let arrow_writer = ArrowWriter::try_new(cursor.clone(), arrow_schema_ref.clone(), Some(writer_properties.clone())).unwrap();
+
+        Ok(Self {
+            table_path,
+            storage,
+            arrow_schema_ref,
+            writer_properties,
+            cursor,
+            arrow_writer,
+            partition_values: HashMap::new(),
+            num_records: 0,
+        })
+    }
+
+    /// Writes the record batch in-memory and updates internal state accordingly.
+    /// This method buffers the write stream internally so it can be invoked for many record batches and flushed after the appropriate number of bytes has been written.
+    pub async fn write_record_batch(&mut self, partition_cols: &Vec<String>, record_batch: &RecordBatch) -> Result<(), DeltaWriterError> {
+        let partition_values = extract_partition_values(partition_cols, record_batch)?;
+
+        // TODO: Verify that this RecordBatch contains partition values that agree with previous partition values (if any) for the current writer context
```

Comment:
True - we do need to do that check here. Thanks, will add. The comment is about the partition values though rather than the schema. Basically, if we are writing a file for a table partitioned by `creation_date` and all records in the first batch have value `2021-03-01`, then all subsequent batches written to the same file must have the same value.

Reply:
I have added the schema check as recommended.